### PR TITLE
[Draft]Testing Hapigen Output for NI Digital

### PR DESCRIPTION
**Pull request by IPetersNI**
### What does this Pull Request accomplish?
This Pull Request test the output of the hapigen files for NI-Digital. This is paired with the Pull Request in azure located here: https://ni.visualstudio.com/DevCentral/_git/ni-central/pullrequest/361686.

### Why should this Pull Request be merged?
This Pull Request should not be merge as it is a draft request. This is only to use for building and seeing the differences within the generated files/built files.

### What testing has been done?
Testing has been done manually using local builds and updates for metadata and generated files.

### CODE REVIEW

File: source/codegen/metadata/nidigitalpattern/functions.py [hunk 1]
```diff
@@ -323,7 +267,7 @@
     },
     'Close': {
         'cname': 'niDigital_close',
-        'codegen_method': 'public',
+        'codegen_method': 'private',
```
Comment (reckenro) [inline source/codegen/metadata/nidigitalpattern/functions.py:270]: Changed to private. Should be public.

File: source/codegen/metadata/nidigitalpattern/functions.py [hunk 2]
```diff
@@ -134,67 +137,6 @@
...
-                'cppName': 'timeout',
-                'direction': 'in',
-                'grpc_type': 'double',
-                'name': 'timeout',
-                'type': 'ViReal64'
-            }
-        ],
-        'returns': 'ViStatus'
-    },
-    'BurstPatternSynchronized': {
```
Comment (reckenro) [inline source/codegen/metadata/nidigitalpattern/functions.py:177]: Missing method.

File: source/codegen/metadata/nidigitalpattern/functions.py [hunk 3]
```diff
@@ -387,7 +333,8 @@
         ],
         'returns': 'ViStatus'
     },
-    'ConfigureCycleNumberHistoryRAMTrigger': {
```
Comment (reckenro) [inline source/codegen/metadata/nidigitalpattern/functions.py:390]: Missing method

File: source/codegen/metadata/nidigitalpattern/functions.py [hunk 4]
```diff
@@ -397,23 +344,17 @@
...
-                'name': 'pretriggerSamples',
-                'type': 'ViInt32'
+                'grpc_type': 'string',
+                'name': 'siteList',
+                'type': 'ViConstString'
             }
         ],
         'returns': 'ViStatus'
     },
-    'ConfigureDigitalEdgeConditionalJumpTrigger': {
```
Comment (reckenro) [inline source/codegen/metadata/nidigitalpattern/functions.py:416]: Missing method

File: source/codegen/metadata/nidigitalpattern/functions.py [hunk 5]
```diff
@@ -423,31 +364,32 @@
...
-                'name': 'edge',
-                'type': 'ViInt32'
+                'grpc_type': 'double',
+                'name': 'strobeEdge',
+                'type': 'ViReal64'
             }
         ],
         'returns': 'ViStatus'
     },
-    'ConfigureDigitalEdgeStartTrigger': {
```
Comment (reckenro) [inline source/codegen/metadata/nidigitalpattern/functions.py:450]: Missing method.

File: source/codegen/metadata/nidigitalpattern/functions.py [hunk 6]
```diff
@@ -457,43 +399,38 @@
...
                 'direction': 'in',
-                'enum': 'DigitalEdge',
-                'grpc_type': 'sint32',
-                'name': 'edge',
-                'type': 'ViInt32'
-            }
-        ],
-        'returns': 'ViStatus'
-    },
-    'ConfigureFirstFailureHistoryRAMTrigger': {
```
Comment (reckenro) [inline source/codegen/metadata/nidigitalpattern/functions.py:477]: Missing method

File: generated/nidigitalpattern/nidigitalpattern.proto [hunk 7]
```diff
@@ -7,7 +7,7 @@
 syntax = "proto3";
 
 option java_multiple_files = true;
-option java_package = "com.ni.grpc.nidigital";
+option java_package = "com.ni.grpc.digital";
```
Comment (reckenro) [inline generated/nidigitalpattern/nidigitalpattern.proto:10]: Changed java_package.

File: generated/nidigitalpattern/nidigitalpattern.proto [hunk 8]
```diff
@@ -21,29 +21,17 @@ service NiDigital {
   rpc ApplyLevelsAndTiming(ApplyLevelsAndTimingRequest) returns (ApplyLevelsAndTimingResponse);
   rpc ApplyTDROffsets(ApplyTDROffsetsRequest) returns (ApplyTDROffsetsResponse);
   rpc BurstPattern(BurstPatternRequest) returns (BurstPatternResponse);
-  rpc BurstPatternSynchronized(BurstPatternSynchronizedRequest) returns (BurstPatternSynchronizedResponse);
```
Comment (reckenro) [inline generated/nidigitalpattern/nidigitalpattern.proto:24]: I think as a first pass getting all these methods here would clear up what else is missing / different.

File: generated/nidigitalpattern/nidigitalpattern.proto [hunk 9]
```diff
@@ -83,40 +57,24 @@ service NiDigital {
...
-  rpc MapPinToChannel(MapPinToChannelRequest) returns (MapPinToChannelResponse);
-  rpc PPMUConfigureApertureTime(PPMUConfigureApertureTimeRequest) returns (PPMUConfigureApertureTimeResponse);
-  rpc PPMUConfigureCurrentLevel(PPMUConfigureCurrentLevelRequest) returns (PPMUConfigureCurrentLevelResponse);
-  rpc PPMUConfigureCurrentLevelRange(PPMUConfigureCurrentLevelRangeRequest) returns (PPMUConfigureCurrentLevelRangeResponse);
-  rpc PPMUConfigureCurrentLimit(PPMUConfigureCurrentLimitRequest) returns (PPMUConfigureCurrentLimitResponse);
-  rpc PPMUConfigureCurrentLimitRange(PPMUConfigureCurrentLimitRangeRequest) returns (PPMUConfigureCurrentLimitRangeResponse);
-  rpc PPMUConfigureOutputFunction(PPMUConfigureOutputFunctionRequest) returns (PPMUConfigureOutputFunctionResponse);
-  rpc PPMUConfigureVoltageLevel(PPMUConfigureVoltageLevelRequest) returns (PPMUConfigureVoltageLevelResponse);
-  rpc PPMUConfigureVoltageLimits(PPMUConfigureVoltageLimitsRequest) returns (PPMUConfigureVoltageLimitsResponse);
+  rpc LockSession(LockSessionRequest) returns (LockSessionResponse);
```
Comment (reckenro) [inline generated/nidigitalpattern/nidigitalpattern.proto:77]: LockSession and UnlockSession should be marked 'codegen_method': 'private' in hapigen.

File: CMakeLists.txt [hunk 10]
```diff
@@ -464,7 +464,6 @@ set(system_test_runner_sources
   "source/tests/system/nidaqmx_session_tests.cpp"
   "source/tests/system/nidcpower_driver_api_tests.cpp"
   "source/tests/system/nidcpower_session_tests.cpp"
-  "source/tests/system/nidigital_driver_api_tests.cpp"
```
Comment (reckenro) [inline CMakeLists.txt:467]: Can you undo this change now that the tests should be fixed?

File: generated/nidigitalpattern/nidigitalpattern.proto [hunk 11]
```diff
@@ -41,7 +41,7 @@ service NiDigital {
   rpc ConfigureStartLabel(ConfigureStartLabelRequest) returns (ConfigureStartLabelResponse);
   rpc ConfigureTerminationMode(ConfigureTerminationModeRequest) returns (ConfigureTerminationModeResponse);
   rpc ConfigureTimeSetCompareEdgesStrobe(ConfigureTimeSetCompareEdgesStrobeRequest) returns (ConfigureTimeSetCompareEdgesStrobeResponse);
-  rpc ConfigureTimeSetCompareEdgesStrobe2x(ConfigureTimeSetCompareEdgesStrobe2xRequest) returns (ConfigureTimeSetCompareEdgesStrobe2xResponse);
+  rpc ConfigureTimeSetCompareEdgesStrobe2X(ConfigureTimeSetCompareEdgesStrobe2XRequest) returns (ConfigureTimeSetCompareEdgesStrobe2XResponse);
```
Comment (reckenro) [inline generated/nidigitalpattern/nidigitalpattern.proto:44]: Need to override the name in hapigen with 'grpc_name'.

File: source/tests/system/nidigital_driver_api_tests.cpp [hunk 12]
```diff
@@ -168,6 +168,7 @@ class NiDigitalDriverApiTest : public ::testing::Test {
 
   bool get_bool_attribute(const char* channel_name, digital::NiDigitalAttribute attribute)
   {
+    /*
```
Comment (reckenro) [inline source/tests/system/nidigital_driver_api_tests.cpp:171]: Can you undo these changes? I think the tests should be good now since all the functions are there.

File: source/codegen/metadata/nidigitalpattern/config.py [hunk 13]
```diff
@@ -1,31 +1,16 @@
-﻿# -*- coding: utf-8 -*-
-# This file is generated from NI-Digital Pattern Driver API metadata version 19.5.0d7
+# -*- coding: utf-8 -*-
+# This file is generated from NI-Digital Pattern Driver API metadata version 22.8.0d9999
 config = {
     'additional_headers': {
-        'custom/ivi_errors.h': [
```
Comment (reckenro) [inline source/codegen/metadata/nidigitalpattern/config.py:5]: Can you add this back? In hapigen, you can see [here in Azure](https://dev.azure.com/ni/DevCentral/_git/ni-central/pullrequest/357617?_a=files&path=/src/nifgen/niFgenBaseDistribution/fgenAPIMetadata/source/fgenAPIMetadata/grpc_device/global_config/addition.hapigen) for an example

File: source/codegen/metadata/nidigitalpattern/config.py [hunk 14]
```diff
@@ -46,38 +31,9 @@
         }
     },
     'linux_rt_support': False,
-    'metadata_version': '2.0',
-    'module_name': 'nidigitalpattern',
```
Comment (reckenro) [inline source/codegen/metadata/nidigitalpattern/config.py:50]: Can you fix 'module_name' back to nidigital? It'll be the same fix done for namespace_component in the hapigen global_config file(s).

File: CMakeLists.txt [hunk 15]
```diff
@@ -467,8 +467,8 @@ set(system_test_runner_sources
...
   "source/tests/system/nidigital_session_tests.cpp"
+  "source/tests/system/nidigital_driver_api_tests.cpp"
```
Comment (reckenro) [inline CMakeLists.txt:471]: Can you put this back above "source/tests/system/nidigital_session_tests.cpp" like it was before? That'll get rid of this file in the diff.

File: generated/nidigitalpattern/nidigitalpattern.proto (hunk 11 above)
Comment (reckenro) [reply generated/nidigitalpattern/nidigitalpattern.proto:44]: It's still showing up with 2X but I think you have it right in functions.py now. Can you rebuild locally and push those generated file changes up to update this?

Comment (reckenro) [reply generated/nidigitalpattern/nidigitalpattern.proto:44]: These diffs are still showing up here. Can you re-build locally and push up the changes to these files?

File: CMakeLists.txt (hunk 10 above)
Comment (dmondrik) [reply CMakeLists.txt:470]: There are still whitespace differences here.

File: generated/nidigitalpattern/nidigitalpattern.proto (hunk 11 above)
Comment (dmondrik) [reply generated/nidigitalpattern/nidigitalpattern.proto:44]: @IPetersNI the latest still has a diff here. Contact me after lunch if you want me to help you figure it out locally.

File: source/codegen/metadata/nidigitalpattern/nidigital.proto [hunk 16]
```diff
@@ -0,0 +1,1838 @@
...
+  rpc ConfigureHistoryRAMCyclesToAcquire(ConfigureHistoryRAMCyclesToAcquireRequest) returns (ConfigureHistoryRAMCyclesToAcquireResponse);
+  rpc ConfigurePatternBurstSites(ConfigurePatternBurstSitesRequest) returns (ConfigurePatternBurstSitesResponse);
+  rpc ConfigurePatternLabelHistoryRAMTrigger(ConfigurePatternLabelHistoryRAMTriggerRequest) returns (ConfigurePatternLabelHistoryRAMTriggerResponse);
+  rpc ConfigureSoftwareEdgeConditionalJumpTrigger(ConfigureSoftwareEdgeConditionalJumpTriggerRequest) returns (ConfigureSoftwareEdgeConditionalJumpTriggerResponse);
+  rpc ConfigureSoftwareEdgeStartTrigger(ConfigureSoftwareEdgeStartTriggerRequest) returns (ConfigureSoftwareEdgeStartTriggerResponse);
+  rpc ConfigureStartLabel(ConfigureStartLabelRequest) returns (ConfigureStartLabelResponse);
+  rpc ConfigureTerminationMode(ConfigureTerminationModeRequest) returns (ConfigureTerminationModeResponse);
+  rpc ConfigureTimeSetCompareEdgesStrobe(ConfigureTimeSetCompareEdgesStrobeRequest) returns (ConfigureTimeSetCompareEdgesStrobeResponse);
+  rpc ConfigureTimeSetCompareEdgesStrobe2x(ConfigureTimeSetCompareEdgesStrobe2xRequest) returns (ConfigureTimeSetCompareEdgesStrobe2xResponse);
```
Comment (dmondrik) [inline source/codegen/metadata/nidigitalpattern/nidigital.proto:44]: This file doesn't belong in this location. But it appears you may have fixed the case issue.

Comment (reckenro) [reply source/codegen/metadata/nidigitalpattern/nidigital.proto:44]: Yeah, we're not adding the hapigen-generated proto files yet. You can just delete this out of metadata/nidigitalpattern.

It does look like the casing is fixed here though, which makes me think once you rebuild locally and push up it'll be fixed in generated/nidigitalpattern.proto too.

File: source/codegen/metadata/nidigitalpattern/enums.py [hunk 17]
```diff
@@ -113,6 +112,19 @@
             }
         ]
     },
+    'HistoryRamCyclesToAcquire': {
```
Comment (dmondrik) [inline source/codegen/metadata/nidigitalpattern/enums.py:115]: This is not only out of ABC order, it also has a different RAM capitalization.

Comment (dmondrik) [reply source/codegen/metadata/nidigitalpattern/enums.py:115]: Oh that's actually the reason, since the sorting is case sensitive.

File: source/codegen/metadata/nidigitalpattern/enums.py [hunk 18]
```diff
@@ -367,7 +396,8 @@
             }
         ]
     },
-    'TdrEndpointTermination': {
+    'TDREndpointTermination': {
```
Comment (dmondrik) [inline source/codegen/metadata/nidigitalpattern/enums.py:399]: Different TDR capitalization.

File: source/codegen/metadata/nidigitalpattern/enums.py (hunk 17 above)
Comment (dmondrik) [reply source/codegen/metadata/nidigitalpattern/enums.py:115]: It's still a different case, Ram vs RAM.

Comment (reckenro) [reply source/codegen/metadata/nidigitalpattern/enums.py:115]: Yeah, I think it's that the others changed to be RAM instead of Ram. So in grpc_device/enums/enums.hapigen can you add 'name' overrides for 'history ram max samples to acquire per site' and 'history ram trigger type' like you did for 'history ram cycles to acquire'.

**Review by dmondrik — APPROVED** [review]
